### PR TITLE
Java: Add manual overlay annotations & discard predicates

### PR DESCRIPTION
**Pull request by kaspersv**

### CODE REVIEW

File: java/ql/lib/semmle/code/java/Overlay.qll [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/** Holds for files fully extracted in the overlay. */
+overlay[local]
+predicate extractedInOverlay(string file) {
+  isOverlay() and
+  // The incremental Java extractor extracts skeletons without method
+  // bodies for dependencies. To approximate fully extracted files in
+  // the overlay, we restrict attention to files that contain an expression
+  // with an enclosing callable.
+  exists(@expr e | callableEnclosingExpr(e, _) and file = getRawFile(e))
```
Comment (jbj) [inline java/ql/lib/semmle/code/java/Overlay.qll]: @smowton suggested in the internal issue for "Java overlay: Low-level accuracy issues" to use the `numlines` table instead of `callableEnclosingExpr` to identify non-skeleton files.

I'm fine with postponing that change (and its evaluation) to a follow-up PR.

File: java/ql/lib/java.qll [hunk 2]
```diff
@@ -26,6 +26,7 @@ import semmle.code.java.KotlinType
 import semmle.code.java.Member
 import semmle.code.java.Modifier
 import semmle.code.java.Modules
+import semmle.code.java.Overlay
```
Comment (jbj) [inline java/ql/lib/java.qll]: This import will bring four top-level predicates into the global scope, visible in autocomplete everywhere. I think this would be better as a `private import` in each of the files that define a discard predicate.

File: java/ql/lib/semmle/code/java/Javadoc.qll [hunk 3]
```diff
@@ -196,3 +196,15 @@ class KtCommentSection extends @ktcommentsection {
...
   string toString() { result = this.getContent() }
 }
+
+overlay[local]
+private predicate discardableJavadoc(string file, @javadoc d) {
+  not isOverlay() and
+  exists(@member m | file = getRawFile(m) and hasJavadoc(m, d))
+}
+
+/** Discard javadoc entities in files fully extracted in the overlay. */
```
Comment (jbj) [inline java/ql/lib/semmle/code/java/Javadoc.qll]: Shouldn't we also have discarding rules for comments that aren't Javadoc?

Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Javadoc.qll]: Those appear to be modelled as javadoc entities as well using `isNormalComment` and `isEolComment` for `/* ... */` comments and `//` comments, respectively. 

There is also `ktComments` for Kotlin comments and `xmlComments` for XML that will need to be handled separately in the future.

File: java/ql/lib/semmle/code/java/Overlay.qll (hunk 1 above)
Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Overlay.qll]: I've updated the PR to use `numlines` instead of `callableEnclosingExpr` as there are known theoretical corner cases where `numlines` should yield more accurate results. DCA experiments showed no effect on overlay accuracy on 191 real-world test cases ([`numlines`](https://github.com/github/codeql-dca-main/tree/data/kaspersv/auto/main/a51b684/1750671655393/reports#overlay-query-result-tuple-count-accuracy) vs. [`callableEnclosingExpr`](https://github.com/github/codeql-dca-main/tree/data/kaspersv/auto/main/a51b684/1750662390960/reports#overlay-query-result-tuple-count-accuracy)).

**Comment by kaspersv** [comment]
The `Compile all queries` CI check should succeed once the next CLI is released.

**Review by jbj — APPROVED** [review]
I'm happy with this if @aschackmull is happy

File: java/ql/lib/semmle/code/java/Overlay.qll [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+/** Holds for files fully extracted in the overlay. */
+overlay[local]
+predicate extractedInOverlay(string file) {
+  isOverlay() and
+  // numlines is used to restrict attention to fully extracted files and
+  // ignore skeleton extracted files in the overlay
+  exists(@locatable l | numlines(l, _, _, _) and file = getRawFile(l))
+}
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Overlay.qll:40]: How about adding the following:

```suggestion
}

/**
 * A `@locatable` in the base variant that should be discarded if its file is
 * extracted in the overlay variant.
 */
overlay[local]
abstract class DiscardableLocatable extends @locatable {
  string getRawFileInBase() { not isOverlay() and result = getRawFile(this) }
}

overlay[discard_entity]
private predicate discardLocatable(@locatable el) {
  extractedInOverlay(el.(DiscardableLocatable).getRawFileInBase())
}
```
I think that can reduce a lot of boilerplate.

File: java/ql/lib/semmle/code/java/Expr.qll [hunk 5]
```diff
@@ -2701,3 +2702,15 @@ class RecordPatternExpr extends Expr, @recordpatternexpr {
...
+private predicate discardableExpr(string file, @expr e) {
+  not isOverlay() and
+  file = getRawFile(e)
+}
+
+/** Discard base expressions in files fully extracted in the overlay. */
+overlay[discard_entity]
+private predicate discardExpr(@expr e) {
+  exists(string file | discardableExpr(file, e) and extractedInOverlay(file))
+}
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Expr.qll]: With the suggested `DiscardableLocatable`, this could simply become:
```suggestion
overlay[local]
private class DiscardableExpr extends DiscardableLocatable, @expr { }
```
I'm unsure whether `overlay[local]` is needed here or whether it's inherited from `DiscardableLocatable`.

File: java/ql/lib/semmle/code/java/Javadoc.qll [hunk 6]
```diff
@@ -196,3 +197,15 @@ class KtCommentSection extends @ktcommentsection {
   /** Gets the string representation of this section. */
   string toString() { result = this.getContent() }
 }
+
+overlay[local]
+private predicate discardableJavadoc(string file, @javadoc d) {
+  not isOverlay() and
+  exists(@member m | file = getRawFile(m) and hasJavadoc(m, d))
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Javadoc.qll]: I believe `@javadoc` is a `@locatable`, so the indirection through `hasJavadoc` is unnecessary, I think. Also, with the abstract class this would simply be
```
private class DiscardableJavadoc extends DiscardableLocatable, @javadoc { }
```

File: java/ql/lib/semmle/code/java/Overlay.qll (hunk 4 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/Overlay.qll:40]: Similarly, for things like `@method` we can use:
```
/**
 * A `@locatable` in the base variant that should be discarded if its file is
 * extracted in the overlay variant and it is itself not extracted in the
 * overlay, that is, it is deleted in the overlay.
 */
overlay[local]
abstract class DiscardableReferableLocatable extends @locatable {
  string getRawFileInBase() { not isOverlay() and result = getRawFile(this) }

  predicate existsInOverlay() { isOverlay() and exists(this) }
}

overlay[discard_entity]
private predicate discardReferableLocatable(@locatable el) {
  exists(DiscardableReferableLocatable drl | drl = el |
    extractedInOverlay(drl.getRawFileInBase()) and
    not drl.existsInOverlay()
  )
}
```

Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Overlay.qll:40]: Nice. I've added the suggested abstraction for locatables in commit 4. I've left the locations case as is, as there is only one discard predicate for locations currently.

File: java/ql/lib/semmle/code/java/Expr.qll (hunk 5 above)
Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Expr.qll]: `overlay[local]` is not inherited from `DiscardableLocatable`, but the `overlay[local?]` annotation from the top-level module declaration is inherited, so `overlay[local]` isn't strictly needed, but does clarify the intend.

File: java/ql/lib/semmle/code/java/Javadoc.qll (hunk 6 above)
Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Javadoc.qll]: Good point; the indirection is unnecessary.

File: java/ql/lib/semmle/code/java/Member.qll [hunk 7]
```diff
@@ -906,31 +906,12 @@ class ExtensionMethod extends Method {
...
-  exists(string file | discardableMethod(file, m) and extractedInOverlay(file))
-}
-
-overlay[local]
-private predicate discardableBaseMethod(string file, @method m) {
-  not isOverlay() and
-  file = getRawFile(m)
+private class DiscardableAnonymousMethod extends DiscardableLocatable, @method {
+  DiscardableAnonymousMethod() {
+    not isOverlay() and
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Member.qll]: This line can be deleted - it's included in `getRawFileInBase()`.

```suggestion
```

Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Member.qll]: Fixed

File: java/ql/lib/semmle/code/java/Overlay.qll [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  string toString() { none() }
+}
+
+overlay[discard_entity]
+private predicate discardLocatable(@locatable el) {
+  extractedInOverlay(el.(DiscardableLocatable).getRawFileInBase())
+}
+
+/**
+ * A `@locatable` in the base variant that should be discarded if its file is
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Overlay.qll]: Minor nit: the locatable is in both variants.
```suggestion
 * A `@locatable` that should be discarded in the base variant if its file is
```

File: java/ql/lib/semmle/code/java/Overlay.qll [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+overlay[local]
+predicate extractedInOverlay(string file) {
+  isOverlay() and
+  // numlines is used to restrict attention to fully extracted files and
+  // ignore skeleton extracted files in the overlay
+  exists(@locatable l | numlines(l, _, _, _) and file = getRawFile(l))
+}
+
+/**
+ * A `@locatable` in the base variant that should be discarded if its file is
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/Overlay.qll]: Same.

```suggestion
 * A `@locatable` that should be discarded in the base variant if its file is
```

File: java/ql/lib/semmle/code/java/Overlay.qll (hunk 8 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/Overlay.qll]: This minor qldoc hiccup was my fault - I shuffled things around while writing it.

Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Overlay.qll]: Fixed

File: java/ql/lib/semmle/code/java/Overlay.qll (hunk 9 above)
Comment (kaspersv) [reply java/ql/lib/semmle/code/java/Overlay.qll]: Fixed

**Review by aschackmull — DISMISSED** [review]
LGTM

**Comment by kaspersv** [comment]
@ginsbach I've merged in main and added a missing QLDoc since @aschackmull's review.
